baselines/common/mpi_adam_optimizer.py

```diff
@@ -1,54 +1,46 @@
 import numpy as np
 import tensorflow as tf
-from baselines.common import tf_util as U
 from baselines.common.tests.test_with_mpi import with_mpi
-from baselines import logger
 try:
     from mpi4py import MPI
 except ImportError:
     MPI = None
 
-class MpiAdamOptimizer(tf.train.AdamOptimizer):
+class MpiAdamOptimizer(tf.Module):
     """Adam optimizer that averages gradients across mpi processes."""
-    def __init__(self, comm, grad_clip=None, mpi_rank_weight=1, **kwargs):
+    def __init__(self, comm, var_list):
+        self.var_list = var_list
         self.comm = comm
-        self.grad_clip = grad_clip
-        self.mpi_rank_weight = mpi_rank_weight
-        tf.train.AdamOptimizer.__init__(self, **kwargs)
-    def compute_gradients(self, loss, var_list, **kwargs):
-        grads_and_vars = tf.train.AdamOptimizer.compute_gradients(self, loss, var_list, **kwargs)
-        grads_and_vars = [(g, v) for g, v in grads_and_vars if g is not None]
-        flat_grad = tf.concat([tf.reshape(g, (-1,)) for g, v in grads_and_vars], axis=0) * self.mpi_rank_weight
-        shapes = [v.shape.as_list() for g, v in grads_and_vars]
-        sizes = [int(np.prod(s)) for s in shapes]
+        self.beta1 = 0.9
```

Comment:
what's the purpose of reimplementing adam update logic here (instead of inheriting it from tf.train.AdamOptimizer)?

Reply:
Great question -- I reimplemented so that I don't specifically inherit from tf Adam, the reason is that I need to wrap everything under tf.function, but MPI.Allreduce will only work with numpy, not symbolic tf tensors.